### DISCUSSION TITLE
ProductController::assignAttributesCombinations() with 5k combinations from 23s to less than 2s with two simple optimizations

### DISCUSSION THREAD

**Discussion by the-ge** [PrestaShop/PrestaShop#36004] [Ideas]
### The problem

Working on a store having products with thousands of variants, I arrived to the `assignAttributesCombinations()` method of ProductController, which took ~ 23 seconds to run.
I found two optimizations that cut the time to less than 2 seconds.
While most of the stores do not have that amount of product variants, maybe these optimizations are still of interest ¯\\\__(ツ)\__/¯

### The solution

In the 8.1 branch, the code I'm talking about is:
```php
889            foreach ($attributes_combinations as &$ac) {
890                foreach ($ac as &$val) {
891                    $val = str_replace(Configuration::get('PS_ATTRIBUTE_ANCHOR_SEPARATOR'), '_', Tools::str2url(str_replace([',', '.'], '-', $val)));
892                }
893            }
```
First optimization (lines 891-893 below) - skipping the processing done in line 894 for empty strings - cut the time from 23 seconds to 13 seconds:
```php
889            foreach ($attributes_combinations as &$ac) {
890                foreach ($ac as &$val) {
891                    if ($val === '') {
892                        continue;
893                    }
894                    $val = str_replace(Configuration::get('PS_ATTRIBUTE_ANCHOR_SEPARATOR'), '_', Tools::str2url(str_replace([',', '.'], '-', $val)));
895                }
896            }
```
The second optimization - taking the class method calls/object creation out of the loop in lines 889-903 - cut the time to less than 2 seconds:
```php
889            $separator = Configuration::get('PS_ATTRIBUTE_ANCHOR_SEPARATOR');
890            $modifier = new StringModifier();
891            $normalize = function(&$value) use ($separator, $modifier) {
892                $value = str_replace(
893                        $separator,
894                        '_',
895                        $modifier->str2url(
896                            str_replace(
897                                [',', '.'],
898                                '-',
899                                $value
900                            )
901                        )
902                    );
903            };
904            foreach ($attributes_combinations as &$ac) {
905                foreach ($ac as &$val) {
906                    if ($val === '') {
907                        continue;
908                    }
909                    $normalize($val);
910                }
911            }
```

### Alternatives

The second optimization could be done in a new method instead of an anonymous function and/or without calling the value by reference.

### Additional context

The code above is from ProductController in the 8.1 branch. The site I'm working and testing on is based on 8.1.2.

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by ChillCode** [comment]
Hi @the-ge 

The change on the str2url method introduced these issues, but already reported and "solved":

https://github.com/PrestaShop/PrestaShop/issues/34453

https://github.com/PrestaShop/PrestaShop/pull/34458

**Comment by the-ge** [comment]
Oh, right. I apologize for not checking properly beforehand. It seems that the changes found their place in 8.1.3.
I replaced ProductController.php and Tools.php in my 8.1.2 with the ones in 8.1.5 and got the big speed improvement.

Still, a marginal improvement  - from 1.8...2seconds to less than 1.5 seconds for 5000 combinations - is brought by enveloping the string processing in a empty string checking condition, going from :
```php
864                foreach ($ac as &$val) {
865                    $val = str_replace(Configuration::get('PS_ATTRIBUTE_ANCHOR_SEPARATOR'), '_', Tools::str2url(str_replace([',', '.'], '-', $val)));
866                }
```
in ProductController, to:
```php
864                foreach ($ac as &$val) {
865                    if ($val !== '') {
866                        $val = str_replace(Configuration::get('PS_ATTRIBUTE_ANCHOR_SEPARATOR'), '_', Tools::str2url(str_replace([',', '.'], '-', $val)));
867                    }
868                }
```

On another totally unrelated matter, English is not my native language, thus not all its subtleties make it to me. Why did you quote `"solved"` in your reply? Thank you!

**Comment by ChillCode** [comment]
You don't need to apologize, just posted cause I remembered the issue with str2url, and as you can see all your quoted lines contain this method.

And "solved" because you're right, passing empty strings to a function that doesn't check for them is a waste of resources and time.

Can you create a PR?

**Comment by the-ge** [comment]
Here you are: https://github.com/PrestaShop/PrestaShop/pull/36039.
